language/tools/move-package/src/resolution/dependency_graph.rs

```diff
@@ -969,12 +1045,183 @@ fn pkg_deps_equal<'a>(
+        BTreeSet::from_iter(other_graph.edges(pkg_name).filter_map(|(_, pkg, dep)| {
+            if dep.finalized {
+                Some((dep, pkg))
+            } else {
+                None
+            }
+        }));
 
     let (pkg_deps, other_deps): (Vec<_>, Vec<_>) = pkg_edges
         .symmetric_difference(&other_edges)
         .partition(|dep| pkg_edges.contains(dep));
     (pkg_deps, other_deps)
 }
+
+/// In presence of dependency overrides, returns a "winning" package if any (in which case it
+/// returns an error). The algorithm relies on the fact that this check is performed when a new
+/// package is added to the graph already containing package with the same name - this allows us to
+/// resolve the conflicts pair-wise even if ultimately a package with a given name is a dependency
+/// of more than two other packages.
+///
+/// First, we try to locate the earliest override dominating a package in question (and bail if not
+/// found). Then we traverse dependency graph starting with the override to see if we reach the
+/// conflicting package through more than one edge. If conflicting package is indeed reachable via
+/// more than one edge than the override is insufficient to resolve conflict (and reported as such),
+/// otherwise the "winning" package is returned.
+///
+/// This only works because when trying to add a conflicting package through a new dependency edge,
+/// we add this edge early in an non-finalized state (before the package is fully processed),
+/// finalizing the edges only after package processing is done.
+fn override_pkg(
+    graphmap: DiGraphMap<PM::PackageName, Dependency>,
+    root_pkg_name: PM::PackageName,
+    pkg_name: PM::PackageName,
+    existing_pkg: &Package,
+    new_pkg: &Package,
+) -> Result<Package> {
+    let root_node: NodeIndex<u32> = node_index(graphmap.to_index(root_pkg_name));
+    let pkg_node: NodeIndex<u32> = node_index(graphmap.to_index(pkg_name));
+    let graph = graphmap.into_graph();
+    if let Some(override_node) = find_override(&graph, root_node, pkg_node)? {
+        let mut pkg_found = vec![];
+        find_override_reachable_pkgs(
+            &graph,
+            override_node,
+            override_node,
+            &pkg_name,
+            existing_pkg,
+            new_pkg,
+            &mut pkg_found,
+        )?;
+        if pkg_found.len() == 1 {
+            return Ok(pkg_found[0].clone());
+        }
+    }
+    bail!(
+        "Conflicting dependencies found:\n{0} = {1}\n{0} = {2}",
+        pkg_name,
+        PackageWithResolverTOML(existing_pkg),
+        PackageWithResolverTOML(new_pkg),
+    );
+}
+
+/// Find the closest valid (dominating) override for a given package.
+fn find_override(
+    graph: &Graph<Symbol, Dependency>,
+    root_node: NodeIndex<u32>,
+    pkg_node: NodeIndex<u32>,
+) -> Result<Option<NodeIndex>> {
+    let all_dominators = algo::dominators::simple_fast(&graph, root_node);
```

Comment:
This is going to result in calculating a lot of dominator trees.  Calculating dominators is in the best case almost linear time, but I think the number of times this gets called makes the overall cost in the algorithm quadratic, which I think is avoidable.

Reply:
Document for future cleanup.

Reply:
This also means you probably don't need to make these free functions -- you can put them in the `impl` block for `DependencyGraph` and always refer to `&self.package_graph`.